controller/src/main/java/io/pravega/controller/server/rentention/AutoRetentionBucketService.java

```diff
@@ -0,0 +1,133 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.controller.server.rentention;
+
+import com.google.common.util.concurrent.AbstractService;
+import io.pravega.client.stream.Stream;
+import io.pravega.client.stream.impl.StreamImpl;
+import io.pravega.common.Exceptions;
+import io.pravega.common.concurrent.Futures;
+import io.pravega.controller.store.stream.OperationContext;
+import io.pravega.controller.store.stream.StreamMetadataStore;
+import io.pravega.controller.task.Stream.StreamMetadataTasks;
+import io.pravega.controller.util.Config;
+import io.pravega.controller.util.RetryHelper;
+import lombok.extern.slf4j.Slf4j;
+
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+
+@Slf4j
+public class AutoRetentionBucketService extends AbstractService implements BucketChangeListener {
+
+    private final int bucketId;
+    private final StreamMetadataStore streamMetadataStore;
+    private final StreamMetadataTasks streamMetadataTasks;
+    private final ScheduledExecutorService executor;
+    private final ConcurrentMap<Stream, CompletableFuture> retentionFutureMap;
+    private final LinkedBlockingQueue<BucketChangeListener.StreamNotification> notifications;
+    private final AtomicBoolean stop;
+
+    public AutoRetentionBucketService(int bucketId, StreamMetadataStore streamMetadataStore,
+                                      StreamMetadataTasks streamMetadataTasks, ScheduledExecutorService executor) {
+        this.bucketId = bucketId;
+        this.streamMetadataStore = streamMetadataStore;
+        this.streamMetadataTasks = streamMetadataTasks;
+        this.executor = executor;
+
+        this.notifications = new LinkedBlockingQueue<>();
+        this.stop = new AtomicBoolean(false);
+        this.retentionFutureMap = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    protected void doStart() {
+        Futures.await(RetryHelper.withIndefiniteRetriesAsync(() -> streamMetadataStore.getStreamsForBucket(bucketId, executor)
```

Comment:
This works counter to how doStart should be. You need to execute something very quickly, then run your startup asynchronously, then in your async code, call notifyStarted/notifyFailed. That will prevent you from eating too many threads while starting.

Reply:
i am removing futures.await here. Does that work?
RetryHelper.withIndefiniteRetriesAsync gives a future and on that i am adding whenComplete(notifyStarted/notifyFailed)